checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+        if inc_read == 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Delimiter was not found",
+            ));
+        }
+        bytes_read += inc_read;
+        if bytes_read >= delim.len() && &buf[(buf.len() - delim.len())..buf.len()] == delim {
+            return Ok(bytes_read);
+        }
+    }
+}
+
+enum CheckerErrorUnfilled<T: std::fmt::Debug + 'static> {
+    Mumble,
+    Offline,
+    InternalError(T),
+}
+
+impl From<io::Error> for CheckerErrorUnfilled<io::Error> {
+    fn from(e: std::io::Error) -> CheckerErrorUnfilled<std::io::Error> {
+        match e.kind() {
+            io::ErrorKind::ConnectionRefused
+            | io::ErrorKind::ConnectionAborted
+            | io::ErrorKind::ConnectionReset => CheckerErrorUnfilled::Offline,
+            io::ErrorKind::UnexpectedEof => CheckerErrorUnfilled::Mumble,
+            _ => CheckerErrorUnfilled::InternalError(e),
+        }
+    }
+}
+
+impl<T: std::fmt::Debug + 'static> CheckerErrorUnfilled<T> {
+    fn with_message(self, err_msg: &'static str) -> CheckerError {
+        match self {
+            CheckerErrorUnfilled::Offline => CheckerError::Offline(err_msg),
+            CheckerErrorUnfilled::Mumble => CheckerError::Mumble(err_msg),
+            CheckerErrorUnfilled::InternalError(err) => {
+                error!("Internal error: {:?}", err);
+                CheckerError::InternalError("Internal Error")
+            }
+        }
+    }
+}
+
+pub fn bytes_debug_repr(bytes: &[u8]) -> String {
+    let mut bytes_repr = "b'".to_owned();
+    for b in bytes {
+        let b_escaped: Vec<_> = std::ascii::escape_default(*b).collect();
+        bytes_repr.push_str(std::str::from_utf8(&b_escaped).unwrap());
+    }
+    bytes_repr.push('\'');
+    bytes_repr
+}
+
+#[derive(Debug)]
+pub struct NotebookClient {
+    pub conn: BufStream<TcpStream>,
+    pub user: Option<NotebookUser>,
+}
+
+const PROMPT: &[u8] = b"\n> ";
+
+impl NotebookClient {
+    const SERVICE_PORT: u16 = 2323;
+
+    #[instrument("CONNECT")]
+    pub async fn connect(request: &CheckerRequest) -> CheckerResult<Self> {
+        info!("Connecting to service");
+        let mut conn =
+            match TcpStream::connect(format!("{}:{}", request.address, Self::SERVICE_PORT)).await {
+                Ok(conn) => BufStream::new(conn),
+                Err(e) => {
+                    info!("Failed to connect to service!, {:?}", e);
+                    return Err(CheckerError::Offline("Connection to service failed"));
+                }
+            };
+
+        debug!("Fetching Welcome Banner");
+        let mut welcome_banner = Vec::with_capacity(256);
+        let bytes_read = match read_until_slice(&mut conn, PROMPT, &mut welcome_banner).await {
+            Err(_) | Ok(0) => Err(CheckerError::Mumble("Failed to fetch welcome banner")),
+            Ok(r) => Ok(r),
+        }?;
+
+        if &welcome_banner[(bytes_read - 3)..bytes_read] != b"\n> " {
```

Comment:
You want `slice.ends_with()` here, right?
Also, this will always be `false`, since `PROMT = b"\n> "` -> better add unittets to the `read_until_slice` method instead, and drop this `if`?

Reply:
The standard rust read_until-API also returns if the byte isn't found but instead EOF or similar is hit.
Since i now return UnexpectedEOF this is unneccesary yeah